test/api/user/get.users.id.spec.ts

```diff
@@ -0,0 +1,59 @@
+import { expect, use } from 'chai'
+import chaijsonSchema from 'chai-json-schema'
+import request from 'supertest'
+import { mongo, testServer } from '../../utils/connect'
+import { userSchema } from '../../../src/entity/user'
+import { config } from '../../../src/utils/config'
+import jwt from 'jsonwebtoken'
+import { Db } from 'mongodb'
+import { generateDatabaseUsers } from '../../utils/helpers'
+
+use(chaijsonSchema)
+
+// global db connection
+let database: Db | undefined
+
+;(async () => {
+    if (!mongo.db) database = await mongo.connect()
+})()
+
+const endpoint = { method: 'GET', route: '/users/:id' }
+
+describe(`${endpoint.method}: ${endpoint.route}`, () => {
+    it('should fail to get a user with wrong token', async () => {
+        const users = await generateDatabaseUsers(database, 1)
+        const testUser = { id: users[0]._id, email: users[0].email }
+        const token = jwt.sign(testUser, 'aaaaaaaaa')
+
+        const res = await request(testServer)
+            .get(endpoint.route.replace(':id', testUser.id))
+            .set('Authorization', 'bearer ' + token)
+
+        expect(res.status).to.be.equal(401)
+    })
+
+    it('should fail to get a user with wrong id', async () => {
+        const users = await generateDatabaseUsers(database, 1)
+        const testUser = { id: users[0]._id, email: users[0].email }
+        const token = jwt.sign(testUser, config.jwt.accessTokenSecret, { expiresIn: config.jwt.accessTokenLife })
+
+        const res = await request(testServer)
+            .get(endpoint.route.replace(':id', 'some_bad_id'))
+            .set('Authorization', 'bearer ' + token)
+
+        expect(res.status).to.be.equal(400)
+    })
+
+    it('should get a user by id', async () => {
+        const users = await generateDatabaseUsers(database, 1)
```

Comment:
here the user is created with direct database operations, another way to prepare data is to use the `POST /users` API, can you compare the pros & cons of these two approaches?

Reply:
**1. Creating the test user with direct database operations**
 **Pros:**
 - In the situation in which the `POST /user/:id` endpoint is broken or has an error, it would cause the other tests that depend on this endpoint to fail. Tests should ideally be as independent as possible to ensure that external factors that are not part of the test do not affect the results of the test.
 - Creating the user with direct database operation allows you to insert data that you may not be able to insert with `POST /user/:id` e.g. fields such as `createdAt`, `updatedAt`, `refreshToken`. This gives you greater control of the test since you can insert different kinds of data to test different test cases easily.

**Cons:**

- Direct database operation can fail because of an external factor such as the database connection failing or hanging. However this would have the same effect on using an endpoint to create test users.

**2. Creating a test user with API endpoint**
**Pros:**
- I can only think of one scenario where one may want to do this: In a _multi-step testing workflow_ that has several requests – This is where a series of requests are used to simulate the usage of the API e.g. Register/Login flow. 
For example, a POST request is executed and it creates a user and returns a token in its response. A user is then authenticated with this token on the server. Then we use a PATCH endpoint to update the user, and we again invoke a GET request to validate the new data. Finally, we DELETE that user and use GET again to verify it no longer exists.
This approach can be used to catch certain errors/bugs/usability issues that independent endpoint tests (with direct database creation of resources) may not catch.

**Cons:** 
- As detailed in the Pros of direct database connection.

**Conclusion**

The first approach is preferred but ideally both can be implemented. Independent integration API tests with direct database connection for creating resources run first and when they pass, a multi-step testing workflow can be run against the API e.g A tool such as [Postman Newman ](https://learning.postman.com/docs/running-collections/using-newman-cli/command-line-integration-with-newman/)test runner can be used for this.

Reply:
current test case is depending on a test helper using the internal implementation details, 
how can we keep the data consistent as data created in normal execution interface of the same layer? 
it's likely that we will maintain the same implementation in two places, one in the `POST /users API`, one in the test helper.

can you explain more about test independence? 
what is the unit to test? what is the boundary of the unit that should be independent?
if each test case is a unit, then do not depend on test data in other test cases,
if each interface is a unit, then do not depend on other interface of the same module/layer, 
if each module/layer is a unit, then do not depend on internal implementation details in other modules/layers,
etc.

Reply:
> How can we keep the data consistent as data created in normal execution interface of the same layer

- This could be accomplished with the insertion of captured state fixture documents that capture the correct state of documents in the database _as they should be_ after they have been _correctly_ created/modified by the API. This is what the `generateDbUsers()` function I use in the test cases does. 
- This can also be done by importing the logic of resource creation (e.g `user.createUser()` from the user controller) into the `generateDbUsers()` function to create test users to ensure consistent implementation.

> can you explain more about test independence?
> what is the unit to test? what is the boundary of the unit that should be independent?

With regards to these API tests, each test case should be independent of data from other test cases. These endpoint test cases can therefore be regarded as units in the sense that they should not depend on test data from other test cases.  

The test helper that generates test data for some of these test cases does not depend on data from other test cases. The test cases do however depend on data in another layer (database layer). 
e.g. 
- a test to `GET users/:id` assumes the user already exists in the database, and thus a test helper inserts this user in the correct state before the test runs.
- a test to `PATCH users/:id` assumes the user already exists in the database, and thus a test helper inserts this user in the correct state before the test runs.

Therefore, in this situation, what I mean by test independence is independence from data generated by other test cases (other endpoint tests).